### PR TITLE
AP-2234: Use delegated functions dates on ApplicationProceedingType records

### PR DESCRIPTION
**Pull request by Alex-Swann**
## Use delegated functions dates on ApplicationProceedingType records

This PR does the bulk of the work to enable delegated function dates to be recorded per proceeding type for an application rather than just per application.

- moved delegated function methods on `LegalAidApplication` out to a mixi-in module
- rename `DelegatedFunctionsDateController` to `ConfirmDelegatedFunctionsDateController` (as well as associated paths and routes) to better describe what it does
- ensure that when in non-multiple proceedings mode, the `used_delegated_functions_on` and `used_delegated_functions_reported_on` to the (one and only) application_proceeding_type record
- new `UsedMultipleDelegatedFunctionsController` and form for collecting DF dates for each proceeding type when in multiple proceeding mode
- `used_delegated_functions?` method on `LegalAidApplication` rejigged to query all `ApplicationProceedingTypes`
- Added code to `ChancesOfSuccess` to determine the lowest ranking of them for MIS reporting
- Extracted code from the three Delegated Functions Date controllers into a common service DelegatedFunctionDateService
- Changed SubstantiveApplicationDeadlineCalculator to accept a date rather than legal_aid_application (so that we can pass in the earliest df date)
- Remove delegated function dates from legal aid application record
- Major surgery to the legal aid application factory to easily enable applications with proceeding types (with or without delegated functions to be setup). These replace 6 old traits that had overlapping responsibilities, did not play well with other traits and have now been marked as deprecated.

[Link to story](https://dsdmoj.atlassian.net/browse/AP-XXX)

Describe what you did and why.

## Checklist

Before you ask people to review this PR:

- [ ] Tests and rubocop should be passing: `bundle exec rake`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [ ] There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- [ ] The PR description should say what you changed and why, with a link to the JIRA story.
- [ ] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/controllers/providers/confirm_delegated_functions_dates_controller.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      legal_aid_application.save!
+    end
+
+    def substantive_application_deadline
+      return unless earliest_delegated_functions_date
+
+      SubstantiveApplicationDeadlineCalculator.call earliest_delegated_functions_date
+    end
+
+    # def add_delegated_scope_limitations
```
Comment (willc-work) [inline app/controllers/providers/confirm_delegated_functions_dates_controller.rb:60]: delete comments in this file

File: app/controllers/providers/used_delegated_functions_controller.rb [hunk 2]
```diff
@@ -3,62 +3,81 @@ class UsedDelegatedFunctionsController < ProviderBaseController
...
     end
 
-    def remove_delegated_scope_limitations
-      application_proceeding_types.each(&:remove_default_delegated_functions_scope_limitation)
+    def earliest_delegated_functions_reported_date
+      @earliest_delegated_functions_reported_date ||= proceeding_with_earliest_delegated_functions&.used_delegated_functions_reported_on
+    end
+
+    def update_scope_limitations
+      # update_substantive_application_deadline
```
Comment (willc-work) [inline app/controllers/providers/used_delegated_functions_controller.rb:55]: delete comments

File: app/forms/legal_aid_applications/used_delegated_functions_form.rb [hunk 3]
```diff
@@ -1,100 +1,156 @@
...
     end
 
-    # Note that this method is first called by `validates`.
-    # Without that validation, the functionality in this method will not be called before save
-    def used_delegated_functions_on
-      return delete_existing_date unless used_delegated_functions_selected?
-      return @used_delegated_functions_on if @used_delegated_functions_on.present?
-      return if date_fields.blank?
-      return date_fields.input_field_values if date_fields.partially_complete? || date_fields.form_date_invalid?
+    # def proceeding_with_earliest_delegated_functions
```
Comment (willc-work) [inline app/forms/legal_aid_applications/used_delegated_functions_form.rb:49]: delete commented out code

File: app/services/reports/mis/application_detail_csv_line.rb [hunk 4]
```diff
@@ -298,7 +298,7 @@ def opponent_attrs # rubocop:disable Metrics/AbcSize
       end
 
       def merits
-        @line << pretty_success_prospect
+        @line << lowest_prospect_of_success
```
Comment (colinbruce) [inline app/services/reports/mis/application_detail_csv_line.rb:301]: What does this have to do with delegated functions?! 🤔

**Review by willc-work — APPROVED** [review]
